File: modules/fbs-core/api/src/main/kotlin/de/thm/ii/fbs/controller/v2/UserController.kt
package de.thm.ii.fbs.controller.v2

import com.fasterxml.jackson.databind.JsonNode
import com.google.gson.Gson
import de.thm.ii.fbs.model.v2.GlobalRole
import de.thm.ii.fbs.model.v2.security.LegacyToken
import de.thm.ii.fbs.model.v2.security.User
import de.thm.ii.fbs.services.v2.persistence.UserRepository
import de.thm.ii.fbs.utils.v2.annotations.CurrentToken
import de.thm.ii.fbs.utils.v2.exceptions.BadRequestException
import de.thm.ii.fbs.utils.v2.exceptions.ForbiddenException
import de.thm.ii.fbs.utils.v2.exceptions.NotFoundException
import org.springframework.http.MediaType
import org.springframework.web.bind.annotation.*
import javax.servlet.http.HttpServletRequest
import javax.servlet.http.HttpServletResponse

@RestController
@CrossOrigin
@RequestMapping("/api/v1", produces = [MediaType.APPLICATION_JSON_VALUE])
class UserController(
        private val userRepository: UserRepository,
) {
    /**
     * Get all users of the system
     * @param req http request
     * @param res http response
     * @return A list of users
     */
    @GetMapping("users")
    @ResponseBody
    @CurrentToken
    fun getAll(@CurrentToken currentToken: LegacyToken, req: HttpServletRequest, res: HttpServletResponse): List<User> {
        val isDocent = courseRegistrationService.getCoursePrivileges(currentToken.id).exists(e => e . _2 == CourseRole . DOCENT)
        if (currentToken.globalRole == GlobalRole.ADMIN || currentToken.globalRole == GlobalRole.MODERATOR || isDocent) {
            return userRepository.findAll()
        } else {
            throw ForbiddenException()
        }
    }

    /**
     * Get a single user
     * @param uid unique user ide
     * @param req http request
     * @param res http response
     * @return A single user
     */
    @GetMapping("users/{uid}")
    @ResponseBody
    fun getOne(@CurrentToken currentToken: LegacyToken, @PathVariable uid: Int, req: HttpServletRequest, res: HttpServletResponse): User {
        val selfRequest = currentToken.id == uid
        val isDocent = courseRegistrationService.getCoursePrivileges(currentToken.id).exists(e => e . _2 == CourseRole . DOCENT)
        if (currentToken.globalRole == GlobalRole.ADMIN || currentToken.globalRole == GlobalRole.MODERATOR || isDocent || selfRequest) {
            return userRepository.findById(uid).orElse(throw NotFoundException())
        } else {
            throw ForbiddenException()
        }
    }

    /**
     * Update password of user
     * @param uid user identification
     * @param req http request
     * @param res http response
     * @param body Content
     */
    @PutMapping("users/{uid}/passwd", consumes = [MediaType.APPLICATION_JSON_VALUE])
    fun updatePassword(@CurrentToken currentToken: LegacyToken, @PathVariable uid: Int, req: HttpServletRequest
                       , res: HttpServletResponse, @RequestBody body: JsonNode) {
        val password = body["passwd"].asText()
        val passwordRepeat = body["passwdRepeat"].asText()

        if (password.isBlank() || password != passwordRepeat) throw BadRequestException("Malformed Request Body")

        if (currentToken.globalRole == GlobalRole.ADMIN || currentToken.id == uid) {
            userRepository.updateUserPassword(uid, password)
        } else {
            throw ForbiddenException()
        }
    }

    /**
     * Update global role of user
     * @param uid user identification
     * @param req http request
     * @param res http response
     * @param body Content
     */
    @PutMapping("users/{uid}/global-role", consumes = [MediaType.APPLICATION_JSON_VALUE])
    fun updateGlobalRole(@CurrentToken currentToken: LegacyToken, @PathVariable uid: Int, req: HttpServletRequest, res: HttpServletResponse, @RequestBody body: JsonNode) {
        if (body["roleName"] == null) throw BadRequestException("Malformed Request Body")
        val newRole = GlobalRole.parse(body["roleName"].asText())

        if (currentToken.globalRole == GlobalRole.ADMIN || currentToken.id == uid) {
            userRepository.updateUserGlobalRole(uid, newRole)
        } else {
            throw ForbiddenException()
        }
    }

    /**
     * Create a new user
     * @param req http request
     * @param res http response
     * @param body Content
     * @return The created user
     */
    @PostMapping("users", consumes = [MediaType.APPLICATION_JSON_VALUE],
            produces = [MediaType.APPLICATION_JSON_VALUE])
    @ResponseBody
    fun create(@CurrentToken currentToken: LegacyToken, req: HttpServletRequest, res: HttpServletResponse, @RequestBody body: JsonNode): User {
        if (currentToken.globalRole != GlobalRole.ADMIN) {
            throw ForbiddenException()
        }
        try {
            val savedUser = Gson().fromJson(body.toString(), User::class.java)
            return userRepository.save(savedUser)
        } catch (e: Exception) {
            throw BadRequestException("Malformed Request Body")
        }
    }

    /**
     * Delete a user
     * @param uid which user has to be deleted
     * @param req http request
     * @param res http response
     */
    @DeleteMapping("users/{uid}")
    fun delete(@CurrentToken currentToken: LegacyToken, @PathVariable uid: Int, req: HttpServletRequest, res: HttpServletResponse) {
        if (currentToken.globalRole == GlobalRole.ADMIN) {
            userRepository.deleteById(uid)
        } else {
            throw ForbiddenException()
        }
    }
}

File: modules/fbs-core/api/src/main/scala/de/thm/ii/fbs/controller/UserController.scala
package de.thm.ii.fbs.controller

import com.fasterxml.jackson.databind.JsonNode
import de.thm.ii.fbs.controller.exception.{BadRequestException, ResourceNotFoundException}
import de.thm.ii.fbs.model.v2.security.authentication.User
import de.thm.ii.fbs.model.v2.security.authorization.GlobalRole
import de.thm.ii.fbs.security.PermissionEvaluator
import de.thm.ii.fbs.services.security.LocalLoginService
import de.thm.ii.fbs.services.v2.security.authentication.UserService
import de.thm.ii.fbs.util.JsonWrapper.jsonNodeToWrapper
import de.thm.ii.fbs.utils.v2.security.authorization.IsAdmin
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.http.MediaType
import org.springframework.security.access.prepost.PreAuthorize
import org.springframework.web.bind.annotation._

import javax.servlet.http.{HttpServletRequest, HttpServletResponse}
import scala.jdk.CollectionConverters.CollectionHasAsScala

/**
  * UserController defines all routes for /users (insert, delete, update)
  */
@RestController
@CrossOrigin
@RequestMapping(path = Array("/api/v1"), produces = Array(MediaType.APPLICATION_JSON_VALUE))
class UserController {
  @Autowired
  private val userService: UserService = null
  @Autowired
  private val loginService: LocalLoginService = null

  /**
    * Get all users of the system
    *
    * @param req http request
    * @param res http response
    * @return A list of users
    */
  @GetMapping(value = Array("users"))
  @ResponseBody
  @PreAuthorize("hasRole('MODERATOR') || @permissions.hasDocentRole()")
  def getAll(req: HttpServletRequest, res: HttpServletResponse): List[User] =
    userService.getAll(true).asScala.toList

  /**
    * Get a single user
    *
    * @param userId unique user ide
    * @param req    http request
    * @param res    http response
    * @return A single user
    */
  @GetMapping(value = Array("users/{userId}"))
  @ResponseBody
  @PreAuthorize("hasRole('MODERATOR') || @permissions.hasDocentRole() || @permissions.isSelf(#userId)")
  def getOne(@PathVariable userId: Int, req: HttpServletRequest, res: HttpServletResponse): User =
    userService.find(userId) match {
      case u: User => u
      case _ => throw new ResourceNotFoundException()
    }

  /**
    * Update password of user
    *
    * @param userId user identification
    * @param req    http request
    * @param res    http response
    * @param body   Content
    */
  @PutMapping(value = Array("users/{userId}/passwd"), consumes = Array(MediaType.APPLICATION_JSON_VALUE))
  @PreAuthorize("hasRole('ADMIN') || @permissions.isSelf(#userId)")
  def updatePassword(@PathVariable userId: Int, req: HttpServletRequest, res: HttpServletResponse, @RequestBody body: JsonNode): Unit = {
    val password = body.retrive("passwd").asText()
    val passwordRepeat = body.retrive("passwdRepeat").asText()

    if (password.isEmpty || password.get.isBlank || password != passwordRepeat) {
      throw new BadRequestException("Malformed Request Body")
    }
    loginService.upgradePassword(PermissionEvaluator.getUser, password.get)
  }

  /**
    * Update global role of user
    *
    * @param userId user identification
    * @param req    http request
    * @param res    http response
    * @param body   Content
    */
  @PutMapping(value = Array("users/{userId}/global-role"), consumes = Array(MediaType.APPLICATION_JSON_VALUE))
  @IsAdmin
  def updateGlobalRole(@PathVariable userId: Int, req: HttpServletRequest, res: HttpServletResponse, @RequestBody body: JsonNode): Unit = {
    val someRoleName = body.retrive("roleName").asText()

    someRoleName match {
      case Some(roleName) => userService.updateGlobalRoleFor(userId, GlobalRole.parse(roleName))
      case None => throw new BadRequestException("Malformed Request Body")
    }
  }

  /**
    * Create a new user
    *
    * @param req  http request
    * @param res  http response
    * @param body Content
    * @return The created user
    */
  @PostMapping(value = Array("users"), consumes = Array(MediaType.APPLICATION_JSON_VALUE),
    produces = Array(MediaType.APPLICATION_JSON_VALUE))
  @ResponseBody
  @IsAdmin
  def create(req: HttpServletRequest, res: HttpServletResponse, @RequestBody body: JsonNode): User =
    (body.retrive("prename").asText(),
      body.retrive("surname").asText(),
      body.retrive("email").asText(),
      body.retrive("password").asText(),
      body.retrive("username").asText(),
      body.retrive("alias").asText(),
      body.retrive("globalRole").asText()
    ) match {
      case (Some(prename), Some(surname), Some(email), Some(password), Some(username), alias, globalRoleName) =>
        loginService.createUser(
          new User(prename, surname, username, globalRoleName.map(GlobalRole.parse).getOrElse(GlobalRole.USER), email, alias.orNull, null, false, false, null),
          password)
      case _ => throw new BadRequestException("Malformed Request Body")
    }

  /**
    * Delete a user
    *
    * @param userId which user has to be deleted
    * @param req    http request
    * @param res    http response
    */
  @DeleteMapping(value = Array("users/{userId}"))
  @IsAdmin
  def delete(@PathVariable userId: Int, req: HttpServletRequest, res: HttpServletResponse): Unit =
    userService.delete(userId)
}
